stable_baselines3/her/her_replay_buffer.py

```diff
@@ -0,0 +1,332 @@
+        """
+        # Select which episodes to use
+        if online_sampling:
+            episode_indices = np.random.randint(0, self.n_episodes_stored, batch_size)
+            her_indices = np.arange(batch_size)[: int(self.her_ratio * batch_size)]
+        else:
+            assert maybe_vec_env is None, "Transitions must be stored unnormalized in the replay buffer"
+            # Offline sampling: there is only one episode stored
+            episode_length = self.episode_lengths[0]
+            # we sample n_sampled_goal per timestep in the episode (only one is stored).
+            episode_indices = np.tile(0, (episode_length * n_sampled_goal))
+            # we only sample virtual transitions
+            # as real transitions are already stored in the replay buffer
+            her_indices = np.arange(len(episode_indices))
+
+        ep_length = self.episode_lengths[episode_indices]
+
+        if self.goal_selection_strategy == GoalSelectionStrategy.FUTURE:
+            # restrict the sampling domain when ep_length > 1
+            # otherwise filter out the indices
+            her_indices = her_indices[ep_length[her_indices] > 1]
+            ep_length[her_indices] -= 1
+
+        if online_sampling:
+            # Select which transitions to use
+            transitions_indices = np.random.randint(ep_length)
+        else:
+            if her_indices.size == 0:
+                # Episode of one timestep, not enough for using the "future" strategy
+                # no virtual transitions are created in that case
+                return np.empty(0), np.empty(0), np.empty(0), np.empty(0)
+            else:
+                # Repeat every transition index n_sampled_goals times
+                # to sample n_sampled_goal per timestep in the episode (only one is stored).
+                # Now with the corrected episode length when using "future" strategy
+                transitions_indices = np.tile(np.arange(ep_length[0]), n_sampled_goal)
+                episode_indices = episode_indices[transitions_indices]
+                her_indices = np.arange(len(episode_indices))
+
+        # get selected transitions
+        transitions = {key: self.buffer[key][episode_indices, transitions_indices].copy() for key in self.buffer.keys()}
+
+        # sample new desired goals and relabel the transitions
+        new_goals = self.sample_goals(episode_indices, her_indices, transitions_indices)
+        transitions["desired_goal"][her_indices] = new_goals
+
+        # Convert info buffer to numpy array
+        transitions["info"] = np.array(
+            [
+                self.info_buffer[episode_idx][transition_idx]
+                for episode_idx, transition_idx in zip(episode_indices, transitions_indices)
+            ]
+        )
+
+        # Vectorized computation
+        transitions["reward"][her_indices, 0] = self.env.env_method(
+            "compute_reward",
+            transitions["next_achieved_goal"][her_indices, 0],
```

Comment:
why do you use "next_achieved_goal" and not "achieved_goal"?

Reply:
it seems to be the case in baselines implementation: https://github.com/openai/baselines/blob/master/baselines/her/her_sampler.py#L52
but it does not make sense to me...

Reply:
if this is right, we may have to check for the timelimit, to store the proper next obsveration when the episode ends

Reply:
nevermind, this is right but we need to fix the next_obs for the last timestep

Reply:
and which info dict is passed